### DISCUSSION TITLE
Delete overlapping pores?

### DISCUSSION THREAD

**Discussion by ShaneFranklinSoil** [PMEAL/OpenPNM#2912] [Q&A]
I am making voronoi to export for microfluidics. I want the particle size to be around the size of sand grains ~0.5 mm. It seems like to get the voronoi this small it comes along with an overlapping pore error. The error won't go away unless particle size becomes quite large. 

Any thoughts here?

**Comment by jgostick** [comment]
you can try the `relaxation` argument in the Voronoi generator.  If you set this to 2 or 3, the base points will move away from each other so they is room for larger pores.  Without this there will always be pores which are quite close together.

**Reply by jgostick** [reply]
Actually, I'm wrong...the relaxation only makes it so that the Delaunay points can be large:
![Figure_1](https://github.com/PMEAL/OpenPNM/assets/5092049/8fb480f0-88df-4a96-b68f-e1600ca43f1e)

**Reply by jgostick** [reply]
But this is probably what you want actually....large, similarly sized Voronoi cells where the sand grains live, while fluid flows around the grains via the voronoi connections (blue lines above).

**Reply by ShaneFranklinSoil** [reply]
Particle size looks good in your example. Would the overlapping pores now function as much larger macropores? It seems like a fine balance between maintaining intended pore size distribution and controlling particle size. 

Thank you so much for the help and for the software. It is a great contribution!

Shane

**Reply by jgostick** [reply]
I'm not quite sure I follow what you mean by macropores now.  BTW, here is the code to make that: 

``` python
import openpnm as op
pn1 = op.network.Voronoi(points=100, shape=[1, 1, 0], relaxation=4)
h = op.visualization.plot_connections(pn1)
h = op.visualization.plot_coordinates(pn1, ax=h, s=100)
```

**Reply by ShaneFranklinSoil** [reply]
Maybe another way to ask, when I upload these into files to make into microfluidics I have tried using the pores (red) and throats (blue). Fluid flow here is only through throats (blue)?

**Reply by ShaneFranklinSoil** [reply]
By macropores I mean the overlapping red

**Reply by jgostick** [reply]
I was feeling bored today, so spent 30 minutes doing this: 

``` python
import openpnm as op
import numpy as np

net, vor = op._skgraph.generators.voronoi(
    shape=[1, 1, 0],
    points=100,
    relaxation=5,
    reflect=True,
    node_prefix='pore',
    edge_prefix='throat',
)
# add delaunay points
pn = op.network.Network()
pn.update(net)
pn['pore.voronoi'] = True
pts = np.hstack((vor.points[:100], np.zeros([100, 1])))
op.topotools.extend(pn, pore_coords=pts)
pn.add_model(
    propname='pore.diameter',
    model=op.models.geometry.pore_size.largest_sphere,
    domain='pore.voronoi',
)
pn['pore.delaunay'] = False
pn['pore.delaunay'][pn.pores('voronoi', mode='not')] = True
pn['pore.diameter@delaunay'] = 0.1

# %%
h = op.visualization.plot_connections(
    pn,
    throats=pn.throats('all'),
)
h = op.visualization.plot_coordinates(
    pn,
    pores=pn.pores('voronoi'),
    size_by=pn['pore.diameter']**2,
    s=500,
    ax=h,
)
h = op.visualization.plot_coordinates(
    pn,
    pores=pn.pores('delaunay'),
    size_by=pn['pore.diameter']**2,
    s=500,
    c='g',
    ax=h,
)
```

Which produces this:
![Figure_1](https://github.com/PMEAL/OpenPNM/assets/5092049/1b70d700-4ea5-4ef4-a6dc-5132a8882229)

**Reply by jgostick** [reply]
Also, not sure what your final aim is exactly, but we have added this to PoreSpy: 

https://porespy.org/examples/generators/reference/cylindrical_pillars_mesh.html

**Reply by ShaneFranklinSoil** [reply]
You're a legend! Sorry I am new to coding and open pnm. This is incredible

**Reply by jgostick** [reply]
The reason the code above is a bit on the verbose and obscure side is because there is a bug in the DelaunayVoronoiDual generator, which I opened an issue for #2913.  Otherwise you could have it all with just one line.
